File: 2/golang/solution1/main.go
package main

import "fmt"

func main() {
	//var num1 = []int{2, 4, 3}
	//var num2 = []int{5, 6, 4}

	//var num1 = []int{0}
	//var num2 = []int{0}

	var num1 = []int{9, 9, 9, 9, 9, 9, 9}
	var num2 = []int{9, 9, 9, 9}
	l1 := createANumber(num1)
	l2 := createANumber(num2)

	printList(l1)
	fmt.Println("----------------------")
	printList(l2)
	fmt.Println("----------------------")
	sum := addTwoNumbers(l1, l2)
	printList(sum)
}

type ListNode struct {
	Val  int
	Next *ListNode
}

func printList(header *ListNode) {
	var temp = header
	for temp != nil {
		fmt.Printf("%d ", temp.Val)
		temp = temp.Next
	}
	fmt.Println()
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var temp []int
	var flag = 0
	for l1 != nil && l2 != nil {
		sum := l1.Val + l2.Val
		if flag == 1 {
			sum += 1
			flag = 0
		}
		if sum > 9 {
			flag = 1
			temp = append(temp, sum%10)
		} else {
			temp = append(temp, sum)
		}
		l1 = l1.Next
		l2 = l2.Next
	}

	var lastList *ListNode
	if l1 != nil {
		lastList = l1
	} else {
		lastList = l2
	}

	for lastList != nil {
		sum := lastList.Val
		if flag == 1 {
			sum += 1
			flag = 0
		}
		if sum > 9 {
			flag = 1
			temp = append(temp, sum%10)
		} else {
			temp = append(temp, sum)
		}
		lastList = lastList.Next
	}
	if flag == 1 {
		temp = append(temp, 1)
	}

	number := createANumber(temp)
	return number
}

func createANumber(nums []int) *ListNode {
	var header ListNode
	var prev *ListNode
	for i, num := range nums {
		if i == 0 {
			header.Val = num
			header.Next = nil
			prev = &header
		} else {
			var temp = ListNode{
				Val:  num,
				Next: nil,
			}
			prev.Next = &temp
			prev = &temp
		}
	}
	return &header
}

File: 2/c++/solution1/main.cpp
//
// Created by tcy on 2023/3/19.
//
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

#include <iostream>

class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        ListNode *result = nullptr;
        ListNode *rear = result;
        int up = 0;
        while (l1 != nullptr || l2 != nullptr) {
            int sum = 0, r=0;
            if (l1 != nullptr) {
                sum += l1->val;
                l1 = l1->next;
            }
            if (l2 != nullptr) {
                sum += l2->val;
                l2 = l2->next;
            }
            sum += up;
            r = sum % 10;
            up = sum / 10;

            if (result == nullptr) {    // 一开始
                auto *temp = new(ListNode);
                temp->val = r;
                result = temp;
                rear = temp;
                continue;
            } else {
                auto *temp = new(ListNode);
                temp->val = r;
                rear->next = temp;
                rear = rear->next;
            }
        }
        if (up != 0) {
            auto *temp = new(ListNode);
            temp->val = up;
            rear->next = temp;
        }
        return result;
    }
};

ListNode *createList(const int nums[], int n) {
    ListNode *result = nullptr;
    ListNode *rear = nullptr;
    for (int i = 0; i < n; ++i) {
        auto *temp = new(ListNode);
        temp->val = nums[i];
        if (result == nullptr) {
            result = temp;
            rear = result;
        }
        rear->next = temp;
        rear = rear->next;
    }
    return result;
}

void showList(ListNode *num) {
    ListNode *temp = num;
    while (temp != nullptr) {
        std::cout << temp->val << ", ";
        temp = temp->next;
    }
}

int main() {
//    int nums1[] = {2,4,3};
//    int nums2[] = {5,6,4};

    int nums1[] = {9, 9, 9, 9, 9, 9, 9};
    int nums2[] = {9, 9, 9, 9};

    auto n1 = sizeof(nums1) / sizeof(int);
    auto n2 = sizeof(nums2) / sizeof(int);
    auto list1 = createList(nums1, n1);
    auto list2 = createList(nums2, n2);

    Solution solution;
    auto temp = solution.addTwoNumbers(list1, list2);
    showList(temp);

    delete (list1);
    delete (list2);
    delete (temp);
}
